File: iOS/Kalorie/Core/UseCases/SetupDefaultMealsUseCase.swift
//
//  SetupDefaultMealsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation

protocol SetupDefaultMealsUseCaseProtocol {
    func callAsFunction() async throws -> [MealTypeDomain]
}

struct SetupDefaultMealsUseCase: SetupDefaultMealsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws -> [MealTypeDomain] {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        var startMinutes = 5 * 60
        var endMinutes = startMinutes + 3 * 60
        let mealNames = [
            L10n.DefaultMeals.breakfast,
            L10n.DefaultMeals.secondBreakfast,
            L10n.DefaultMeals.lunch,
            L10n.DefaultMeals.snack,
            L10n.DefaultMeals.dinner
        ]
        var dtos: [(item: MealTypeDTO, id: String)] = []
        var domains: [MealTypeDomain] = []

        for mealName in mealNames {
            let id = UUID().uuidString
            dtos.append((
                item: MealTypeDTO(
                    id: id,
                    name: mealName,
                    startMinutes: startMinutes,
                    endMinutes: endMinutes
                ),
                id: id
            ))
            domains.append(MealTypeDomain(id: id, name: mealName, startMinutes: startMinutes, endMinutes: endMinutes))
            startMinutes = endMinutes
            endMinutes = startMinutes + 3 * 60
        }

        try await dataProvider.batchSetAsync(dtos, in: Constants.Firestore.mealTypes(userId: userId))
        return domains
    }
}

#if DEBUG
struct SetupDefaultMealsUseCaseFake: SetupDefaultMealsUseCaseProtocol {

    // MARK: - Properties

    var stubbedTypes: [MealTypeDomain] = []

    // MARK: - Functions

    func callAsFunction() async throws -> [MealTypeDomain] { stubbedTypes }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SetupDefaultMealsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MealTypeDTO
import antoni.kalorie.core.networking.batchSetAsync
import antoni.kalorie.core.utils.Constants
import java.util.UUID

interface SetupDefaultMealsUseCaseProtocol {
    suspend operator fun invoke(): List<MealTypeDomain>
}

class SetupDefaultMealsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
    private val mealNames: List<String>,
) : SetupDefaultMealsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): List<MealTypeDomain> {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        var startMinutes = DEFAULT_START_HOUR * 60
        var endMinutes = startMinutes + DEFAULT_WINDOW_HOURS * 60
        val dtos = mutableListOf<Pair<MealTypeDTO, String>>()
        val domains = mutableListOf<MealTypeDomain>()

        for (mealName in mealNames) {
            val id = UUID.randomUUID().toString().uppercase()
            dtos += MealTypeDTO(
                id = id,
                name = mealName,
                startMinutes = startMinutes,
                endMinutes = endMinutes,
            ) to id
            domains += MealTypeDomain(id = id, name = mealName, startMinutes = startMinutes, endMinutes = endMinutes)
            startMinutes = endMinutes
            endMinutes = startMinutes + DEFAULT_WINDOW_HOURS * 60
        }

        dataProvider.batchSetAsync(dtos, inCollection = Constants.Firestore.mealTypes(userId))
        return domains
    }

    private companion object {
        const val DEFAULT_START_HOUR = 5
        const val DEFAULT_WINDOW_HOURS = 3
    }
}
